### ISSUE TITLE
[GR-51363] AssertionError when using `-J-ea -H:+ExtendedAsserts`

### ISSUE THREAD

**Issue by zakkak** [oracle/graal#7562]
**Describe the issue**
Running `native-image` with `-J-ea -H:+ExtendedAsserts` results in an `AssertionError`

**Steps to reproduce the issue**
```
cat <<EOF > HelloWorld.java                  
public class HelloWorld {
    public static void main(String[] args) {
        System.out.println("Hello World!");
    }
}
EOF
javac HelloWorld.java
native-image -J-ea -H:+ExtendedAsserts HelloWorld
```

**Describe GraalVM and your environment:**
 - GraalVM version : **a22eac89b35be929604b56881e127c200d4c2721**
 - JDK major version: **22**
 - OS: **Fedora 38**
 - Architecture: **AMD64**

**More details**
```
Warning: The option '-H:+ExtendedAsserts' is experimental and must be enabled via '-H:+UnlockExperimentalVMOptions' in the future.
Warning: Please re-evaluate whether any experimental option is required, and either remove or unlock it. The build output lists all active experimental options, including where they come from and possible alternatives. If you think an experimental option should be considered as stable, please file an issue.
========================================================================================================================
GraalVM Native Image: Generating 'helloworld' (executable)...
========================================================================================================================
[1/8] Initializing...                                                                                    (1.4s @ 0.23GB)
 Java version: 22+16, vendor version: GraalVM CE 22-dev+16.1
 Graal compiler: optimization level: 2, target machine: x86-64-v3
 C compiler: gcc (redhat, x86_64, 13.2.1)
 Garbage collector: Serial GC (max heap size: 80% of RAM)
 1 user-specific feature(s):
 - com.oracle.svm.thirdparty.gson.GsonFeature
------------------------------------------------------------------------------------------------------------------------
Build resources:
 - 26.49GB of memory (42.8% of 61.93GB system memory, determined at start)
 - 32 thread(s) (100.0% of 32 available processor(s), determined at start)
Processing a type not yet marked as instantiated: Lcom/oracle/svm/core/jdk/UninterruptibleUtils$AtomicWord;
[2/8] Performing analysis...  []                                                                         (0.2s @ 0.23GB)
      562 reachable types   (38.9% of    1,445 total)
       74 reachable fields  ( 6.8% of    1,084 total)
      833 reachable methods (20.6% of    4,036 total)
      183 types,     0 fields, and    67 methods registered for reflection

Fatal error: java.lang.AssertionError
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.typestate.TypeState.forIntersection(TypeState.java:219)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.InvokeTypeFlow.filterReceiverState(InvokeTypeFlow.java:183)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.typestate.DefaultSpecialInvokeTypeFlow.onObservedUpdate(DefaultSpecialInvokeTypeFlow.java:66)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.flow.TypeFlow.update(TypeFlow.java:620)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.PointsToAnalysis$1.run(PointsToAnalysis.java:491)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.CompletionExecutor.executeCommand(CompletionExecutor.java:187)
	at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.util.CompletionExecutor.lambda$executeService$0(CompletionExecutor.java:171)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
------------------------------------------------------------------------------------------------------------------------
                         0.0s (2.2% of total time) in 7 GCs | Peak RSS: 0.74GB | CPU load: 6.33
========================================================================================================================
```

cc @christianwimmer @cstancu

**Comment by elkorchi** [comment]
Issue reproduced and tracked internally for a future fix

**Comment by cstancu** [comment]
I replicated the issue with the old GraalVM 23.1.1 release and I investigated why the assertion failed, and I am convinced that this is not a bug. The assertion was too restrictive: a constant can be published in the type flow graphs via a `ConstantTypeFlow` before it is fully scanned and its type is marked as instantiated since `ImageHeapScanner.markReachable()` schedules a concurrent task. In the meantime we have removed that assertion anyway. [We also removed the `ExtendedAsserts` option](https://github.com/oracle/graal/pull/8324) and any of the guarded asserts that were worth keeping are now normal `assert`.
